AttributeError: '_OpNamespace' 'torchvision' object has no attribute 'roi_align'

Hi, I'm getting the following error while running the e2e tests in Torch-MLIR. I'm trying to get a small reproducible for the issue. Meanwhile, if someone is familiar with the error or knows what could be the probable reason behind this, it would be helpful to know. Some changes have been made here: https://github.com/pytorch/vision/commit/8f2e5c90ce0f55877eddb1f7fee8f8b48004849b, but I'm not sure how this relates to the below error.

```
Traceback (most recent call last): 

38535 File "/main_checkout/torch-mlir/docker_venv/lib/python3.10/site-packages/torch/_ops.py", line 705, in __getattr__ 

38536 op, overload_names = torch._C._jit_get_operation(qualified_op_name) 

38537RuntimeError: No such operator torchvision::roi_align 

38538 

38539The above exception was the direct cause of the following exception: 

38540 

38541Traceback (most recent call last): 

38542 File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main 

38543 return _run_code(code, main_globals, None, 

38544 File "/usr/lib/python3.10/runpy.py", line 86, in _run_code 

38545 exec(code, run_globals) 

38546 File "/main_checkout/torch-mlir/e2e_testing/main.py", line 34, in <module> 

38547 register_all_tests() 

38548 File "/main_checkout/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/test_suite/__init__.py", line 18, in register_all_tests 

38549 from . import vision_models 

38550 File "/main_checkout/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir/torch_mlir_e2e_test/test_suite/vision_models.py", line 7, in <module> 

38551 import torchvision.models as models 

38552 File "/main_checkout/torch-mlir/docker_venv/lib/python3.10/site-packages/torchvision/__init__.py", line 6, in <module> 

38553 from torchvision import _meta_registrations, datasets, io, models, ops, transforms, utils 

38554 File "/main_checkout/torch-mlir/docker_venv/lib/python3.10/site-packages/torchvision/_meta_registrations.py", line 22, in <module> 

38555 @register_meta(vision.roi_align.default) 

38556 File "/main_checkout/torch-mlir/docker_venv/lib/python3.10/site-packages/torch/_ops.py", line 709, in __getattr__ 

38557 raise AttributeError( 

38558AttributeError: '_OpNamespace' 'torchvision' object has no attribute 'roi_align' 

38559Error: Process completed with exit code 1.
```

--- a/torchvision/_meta_registrations.py
+++ b/torchvision/_meta_registrations.py
@@ -6,43 +6,45 @@ import torchvision.extension  # noqa: F401
 
 from torch._prims_common import check
 
-_meta_lib = torch.library.Library("torchvision", "IMPL", "Meta")
+if torchvision.extension._has_ops():
 
-vision = torch.ops.torchvision
+    _meta_lib = torch.library.Library("torchvision", "IMPL", "Meta")
 
-
-def register_meta(op):
-    def wrapper(fn):
-        _meta_lib.impl(op, fn)
-        return fn
-
-    return wrapper
+    vision = torch.ops.torchvision
 
 
-@register_meta(vision.roi_align.default)
-def meta_roi_align(input, rois, spatial_scale, pooled_height, pooled_width, sampling_ratio, aligned):
-    check(rois.size(1) == 5, lambda: "rois must have shape as Tensor[K, 5]")
-    check(
-        input.dtype == rois.dtype,
-        lambda: (
-            "Expected tensor for input to have the same type as tensor for rois; "
-            f"but type {input.dtype} does not equal {rois.dtype}"
-        ),
-    )
-    num_rois = rois.size(0)
-    _, channels, height, width = input.size()
-    return input.new_empty((num_rois, channels, pooled_height, pooled_width))
+    def register_meta(op):
+        def wrapper(fn):
+            _meta_lib.impl(op, fn)
+            return fn
+
+        return wrapper
 
 
-@register_meta(vision._roi_align_backward.default)
-def meta_roi_align_backward(
-    grad, rois, spatial_scale, pooled_height, pooled_width, batch_size, channels, height, width, sampling_ratio, aligned
-):
-    check(
-        grad.dtype == rois.dtype,
-        lambda: (
-            "Expected tensor for grad to have the same type as tensor for rois; "
-            f"but type {grad.dtype} does not equal {rois.dtype}"
-        ),
-    )
-    return grad.new_empty((batch_size, channels, height, width))
+    @register_meta(vision.roi_align.default)
+    def meta_roi_align(input, rois, spatial_scale, pooled_height, pooled_width, sampling_ratio, aligned):
+        check(rois.size(1) == 5, lambda: "rois must have shape as Tensor[K, 5]")
+        check(
+            input.dtype == rois.dtype,
+            lambda: (
+                "Expected tensor for input to have the same type as tensor for rois; "
+                f"but type {input.dtype} does not equal {rois.dtype}"
+            ),
+        )
+        num_rois = rois.size(0)
+        _, channels, height, width = input.size()
+        return input.new_empty((num_rois, channels, pooled_height, pooled_width))
+
+
+    @register_meta(vision._roi_align_backward.default)
+    def meta_roi_align_backward(
+        grad, rois, spatial_scale, pooled_height, pooled_width, batch_size, channels, height, width, sampling_ratio, aligned
+    ):
+        check(
+            grad.dtype == rois.dtype,
+            lambda: (
+                "Expected tensor for grad to have the same type as tensor for rois; "
+                f"but type {grad.dtype} does not equal {rois.dtype}"
+            ),
+        )
+        return grad.new_empty((batch_size, channels, height, width))

--- a/torchvision/_meta_registrations.py
+++ b/torchvision/_meta_registrations.py
@@ -1,3 +1,5 @@
+import functools
+
 import torch
 import torch.library
 

--- a/torchvision/_meta_registrations.py
+++ b/torchvision/_meta_registrations.py
@@ -6,45 +6,45 @@ import torchvision.extension  # noqa: F401
 
 from torch._prims_common import check
 
-if torchvision.extension._has_ops():
 
-    _meta_lib = torch.library.Library("torchvision", "IMPL", "Meta")
-
-    vision = torch.ops.torchvision
-
-
-    def register_meta(op):
-        def wrapper(fn):
-            _meta_lib.impl(op, fn)
-            return fn
-
-        return wrapper
+@functools.lru_cache(None)
+def get_meta_lib():
+    return torch.library.Library("torchvision", "IMPL", "Meta")
 
 
-    @register_meta(vision.roi_align.default)
-    def meta_roi_align(input, rois, spatial_scale, pooled_height, pooled_width, sampling_ratio, aligned):
-        check(rois.size(1) == 5, lambda: "rois must have shape as Tensor[K, 5]")
-        check(
-            input.dtype == rois.dtype,
-            lambda: (
-                "Expected tensor for input to have the same type as tensor for rois; "
-                f"but type {input.dtype} does not equal {rois.dtype}"
-            ),
-        )
-        num_rois = rois.size(0)
-        _, channels, height, width = input.size()
-        return input.new_empty((num_rois, channels, pooled_height, pooled_width))
+def register_meta(op_name, overload_name="default"):
+    def wrapper(fn):
+        if torchvision.extension._has_ops():
+            get_meta_lib().impl(getattr(getattr(torch.ops.torchvision, op_name), overload_name), fn)
+        return fn
+
+    return wrapper
 
 
-    @register_meta(vision._roi_align_backward.default)
-    def meta_roi_align_backward(
-        grad, rois, spatial_scale, pooled_height, pooled_width, batch_size, channels, height, width, sampling_ratio, aligned
-    ):
-        check(
-            grad.dtype == rois.dtype,
-            lambda: (
-                "Expected tensor for grad to have the same type as tensor for rois; "
-                f"but type {grad.dtype} does not equal {rois.dtype}"
-            ),
-        )
-        return grad.new_empty((batch_size, channels, height, width))
+@register_meta("roi_align")
+def meta_roi_align(input, rois, spatial_scale, pooled_height, pooled_width, sampling_ratio, aligned):
+    check(rois.size(1) == 5, lambda: "rois must have shape as Tensor[K, 5]")
+    check(
+        input.dtype == rois.dtype,
+        lambda: (
+            "Expected tensor for input to have the same type as tensor for rois; "
+            f"but type {input.dtype} does not equal {rois.dtype}"
+        ),
+    )
+    num_rois = rois.size(0)
+    _, channels, height, width = input.size()
+    return input.new_empty((num_rois, channels, pooled_height, pooled_width))
+
+
+@register_meta("_roi_align_backward")
+def meta_roi_align_backward(
+    grad, rois, spatial_scale, pooled_height, pooled_width, batch_size, channels, height, width, sampling_ratio, aligned
+):
+    check(
+        grad.dtype == rois.dtype,
+        lambda: (
+            "Expected tensor for grad to have the same type as tensor for rois; "
+            f"but type {grad.dtype} does not equal {rois.dtype}"
+        ),
+    )
+    return grad.new_empty((batch_size, channels, height, width))

